ADC/Core/Src/main.c

```diff
@@ -0,0 +1,184 @@
+/*
+ * 	Implement a routine that reads the internal voltage reference (ADC_IN17)
+ * 	and internal temperature sensor (ADC_IN18).
+ * 	The internal reference voltage must be in millivolts and the temperature in Celsius.
+ *
+ * 	1 - Polling		[/]
+ * 	2 - Interrupt	[ ]
+ * 	3 - DMA			[ ]
+ */
+
+#include "stm32l053xx.h"
+#include "macros.h"
+#include "init.h"
+
+uint8_t	AD_CAL 		= 0,
+		T_Delay		= 0;
+int		measure 	= 0,
+		temperature	= 0,
+		v_ref 		= 0;
+
+void ADC_Cal (void);
+void ADC_ON (void);
+void ADC_OFF (void);
+
+int main (void)
+{
+	ADC_Init();
+
+	ADC_ON();
+
+	//Reading the temperature sensor
+	if(ADC1->CR & ADC_CR_ADSTART)			//we have to be sure there's no ongoing conversion
+	{
+		ADC1->CR |= ADC_CR_ADSTP;
+		while(ADC1->CR & ADC_CR_ADSTP);
+	}
+	ADC1->CHSELR |= ADC_CHSELR_CHSEL18;		//selecting the TSEN channel
+	ADC->CCR |= ADC_CCR_TSEN; 				//enables temperature sensor
+	T_Delay = 21;							//we have to wait for 10 us for the temp sensor to wake up
+	while(T_Delay)
+	{
+		T_Delay--;
+	}
+	ADC1->CR |= ADC_CR_ADSTART;				//we start the conversion
+	T_Delay = 42;							//and wait for the properly time
+	while(T_Delay)
+	{
+		T_Delay--;
+	}
+	measure = ADC1->DR & ADC_DR_DATA;		//we store the ADC data in measure
+
+	//temp calc
+	temperature = ((measure * VDD_APPLI / VDD_CALIB) - (int32_t) *TEMP30_CAL_ADDR);
+	temperature = temperature * (int32_t)(130 - 30);
+	temperature = temperature / (int32_t)(*TEMP130_CAL_ADDR - *TEMP30_CAL_ADDR);
+	temperature = temperature + 30;
+
+	//Reading the internal reference voltage
+	if(ADC1->CR & ADC_CR_ADSTART)			//we have to be sure there's no ongoing conversion
+	{
+		ADC1->CR |= ADC_CR_ADSTP;
+		while(ADC1->CR & ADC_CR_ADSTP);
+	}
+	ADC1->CHSELR |= ADC_CHSELR_CHSEL17;		//selecting the VREF channel
+	ADC->CCR |= ADC_CCR_VREFEN; 			//enables temperature sensor
+	T_Delay = 21;							//we have to wait for 10 us for the temp sensor to wake up
+	while(T_Delay)
+	{
+		T_Delay--;
+	}
+	ADC1->CR |= ADC_CR_ADSTART;				//we start the conversion
+	T_Delay = 42;							//and wait for the properly time
+	while(T_Delay)
+	{
+		T_Delay--;
+	}
+	measure = ADC1->DR & ADC_DR_DATA;		//we store the ADC data in measure
+
+	//vref calc
+	v_ref = (3000 * (int32_t) (*VREFINT_CAL)) / measure;	//*3000 so we have Vref in mV
+
+	ADC_OFF();
+
+	while(1)
+	{
+
+	}
+
+	return 0;
+}
+
+void ADC_Cal (void)
+{
+	if(ADC1->CR & ADC_CR_ADEN)						//we make sure that ADEN = 0
+	{
+		ADC1->CR |= ADC_CR_ADDIS;
+		while(ADC1->CR & ADC_CR_ADEN);
+	}
+	ADC1->CFGR1 &= ~ADC_CFGR1_DMAEN;				//and DMAEN = 0
+	ADC1->CR |= ADC_CR_ADVREGEN;					//enables the voltage regulator
+	ADC1->CR |= ADC_CR_ADCAL;						//and then we set ADCAL
+	while(ADC1->CR & ADC_CR_ADCAL);					//we wait until ADCAL = 0 (can be handled by interrupt)
+	AD_CAL =(ADC1->CALFACT & ADC_CALFACT_CALFACT);
+}
+
+void ADC_ON (void)
```

Comment:
I used functions to turn it on and off because it's easier, in a bigger firmware to do so. But I can change it

Reply:
But I'm not turning ADC on and off every new read. I turn it on, read what I need (temperature and Vref) and then turn it off. It's just that I'm reading these values once during the process, and not multiple times.

Reply:
Look, the ADC_OFF function makes sense, after you read you disable the ADC, that's good!
But the ADC_ON it's not needed because you should do this inside ADC_config (take a look on this in TDR code), after you configure the channel. The way you are configuring each channel here it isn't good, because if we have many ADC readings you're gonna have to stop the ADC and configure the channel many times inside the main.c and that's why we use the ADC_config function, inside it we stop the ADC, configure the channel according to the function parameter and then we enable the ADC, so we just have to start the ADC in the main. Please take a look on ADC_config and do something like this for this code.